### PR TITLE
fix(java): don't emit unqualified enum default for discriminator ref-to-property

### LINKED ISSUE #24874: [BUG][java] Discriminator default for direct property enum $ref generates String.<value>
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

### Describe the bug

The Java client generator produces Java that does not compile when a discriminator property is a direct `$ref` to an inline enum property. The generated child constructor assigns the discriminator mapping value as though the property were `String`:

```java
this.category = String.ARCHIVE;
```

`String.ARCHIVE` does not exist.

### Minimal OpenAPI 3.1 input

```yaml
openapi: 3.1.0
info:
  title: Shared category discriminator
  version: 1.0.0
paths: {}
components:
  schemas:
    CategorySource:
      type: object
      required:
        - category
      properties:
        category:
          type: string
          enum:
            - ARCHIVE
    CategoryEvent:
      type: object
      required:
        - category
      properties:
        category:
          description: Event category
          $ref: "#/components/schemas/CategorySource/properties/category"
      discriminator:
        propertyName: category
        mapping:
          ARCHIVE: "#/components/schemas/ArchiveCategoryEvent"
    ArchiveCategoryEvent:
      allOf:
        - $ref: "#/components/schemas/CategoryEvent"
        - type: object
          properties:
            value:
              type: string
```

### Reproduction

```bash
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar \
  generate \
  -g java \
  -i enum-discriminator.yaml \
  -o generated \
  --skip-validate-spec \
  --additional-properties modelNamePrefix=Stock,hideGenerationTimestamp=true

cd generated
mvn -q -DskipTests compile
```

### Actual result

Latest master (`b391b2b98c5926ee9842ef3bc863b57c4d824bbe`, `7.26.0-SNAPSHOT`) generates:

```java
public StockArchiveCategoryEvent() {
    this.category = String.ARCHIVE;
}
```

Compilation fails:

```text
error: cannot find symbol
  symbol:   variable ARCHIVE
  location: class String
```

The same failure occurs with OpenAPI Generator 7.24.0.

### Expected result

The direct property reference should retain the enum type at the discriminator use site. The generated constructor should assign the enum member, rather than `String.ARCHIVE`, and the generated client should compile.

### PR DESCRIPTION
**Pull request by seonwooj0810**
Fixes #24874

## Bug
The Java client generator (default library: okhttp-gson) produced non-compiling code when a discriminator property is a `$ref` directly into another schema's inline enum property, rather than a `$ref` to a named enum schema:

```java
this.category = String.ARCHIVE;
```

`String.ARCHIVE` does not exist, so the generated client fails to compile.

## Root cause
`DefaultCodegen#updateCodegenPropertyEnum` resolves the enum default value and, whenever a matching enum entry is found, unconditionally calls `toEnumDefaultValue(var, enumName)` to qualify it with the property's datatype. That qualification is only meaningful when the property is itself an inline enum (`var.isEnum`) or resolves to a named enum schema (`referencedSchema`). For a discriminator property that is a `$ref` into another schema's *property* (not the enum schema itself), neither is true — `var.datatypeWithEnum` is just the raw scalar type (`String`) — yet the enum-matching logic still ran and produced a bogus qualified token.

This is exercised by okhttp-gson's `pojo.mustache`, which assigns the discriminator's default value in the generated subtype's constructor (other Java libraries don't emit this constructor-level assignment, which is why @jpfinne's comment on the issue pinned it to okhttp-gson specifically — the underlying bad value is computed generically in `DefaultCodegen`, okhttp-gson is just the template that surfaces it as a compile error).

## Fix
Only call `toEnumDefaultValue` when `var.isEnum || referencedSchema.isPresent()`. Otherwise, drop the default (`null`) rather than emit an unqualified token that won't compile. The field is simply left unset in this edge case, which is the same as any other discriminator property whose caller must set it explicitly — no compile error, no incorrect value.

## Test evidence
- Added `modules/openapi-generator/src/test/resources/bugs/issue_24874.yaml` (the exact spec from the issue).
- Added `JavaClientCodegenTest#testDiscriminatorPropertyRefToEnumDoesNotEmitInvalidDefault_issue24874`, which fails on current master (reproduces `this.category = String.ARCHIVE;` via `fileDoesNotContain` assertion) and passes with this fix.
- Ran the full `java-*` sample set (`bin/configs/java-*.yaml`, 136 generators via `bin/generate-samples.sh`) — zero model/pojo diffs, confirming no existing committed sample exercises this narrow edge case and this change is safe.

## Verification done
1. No in-flight PR/linked branch for #24874 (checked via `gh issue view` / `gh api .../timeline`).
2. No active assignee/self-claim.
3. Code-focused change (.java + test fixture only).
4. Confirmed the bug still reproduces on latest `upstream/master` before the fix, and is resolved after.
5. N/A (no parent epic).
6. N/A (not spring-projects/*, no triage-pending gate).

Verification done: reproduced the exact compile failure from the issue on latest master using the issue's own minimal spec, confirmed root cause in `DefaultCodegen#updateCodegenPropertyEnum`, added a regression test that is red before / green after the fix, and confirmed zero sample diffs across the full java-* config set.

---
## Summary by cubic
Fixes #24874. Stops the Java generator from emitting uncompilable defaults like `this.category = String.ARCHIVE;` in okhttp-gson when a discriminator property is a `$ref` to another schema's inline enum property. The default is now left unset instead of qualified with the raw scalar type.

- `toEnumDefaultValue` is now skipped for properties that aren't inline enums and whose datatype is a language-specific primitive; those properties get `null`.
- The guard checks `languageSpecificPrimitives` rather than `referencedSchema`, so package-qualified datatypes (for example PHP) keep valid defaults for named enum refs.
- Adds a regression test using the issue's spec; it failed before this change.
- The full `java-*` sample set regenerates with zero diffs, so no existing sample is affected.

<sup>Written for commit 77718a38a3acc9b222bb7f9421b46166de16d485. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/OpenAPITools/openapi-generator/pull/24934?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by wing328** [comment]
thanks for the pr

please review the build failure when you've time

```
[ERROR] Failures:
[ERROR] org.openapitools.codegen.php.PhpSymfonyServerCodegenTest.testModelPropertyEnumRefWithDefaultUsesNativeEnumCaseInCodegen
[ERROR]   Run 1: PhpSymfonyServerCodegenTest.testModelPropertyEnumRefWithDefaultUsesNativeEnumCaseInCodegen:430 Enum ref + OpenAPI default should set CodegenProperty.defaultValue (see AbstractPhpCodegen.toDefaultValue ref+default handling and updateCodegenPropertyEnum) expected object to not be null
[ERROR]   Run 2: PhpSymfonyServerCodegenTest.testModelPropertyEnumRefWithDefaultUsesNativeEnumCaseInCodegen:430 Enum ref + OpenAPI default should set CodegenProperty.defaultValue (see AbstractPhpCodegen.toDefaultValue ref+default handling and updateCodegenPropertyEnum) expected object to not be null
[ERROR]   Run 3: PhpSymfonyServerCodegenTest.testModelPropertyEnumRefWithDefaultUsesNativeEnumCaseInCodegen:430 Enum ref + OpenAPI default should set CodegenProperty.defaultValue (see AbstractPhpCodegen.toDefaultValue ref+default handling and updateCodegenPropertyEnum) expected object to not be null

```

cc @OpenAPITools/generator-core-team

**Comment by seonwooj0810** [comment]
Thanks for the heads up — the PHP Symfony test failure was caused by the `referencedSchema.isPresent()` check in the else-branch guard: for PHP, the resolved model name is namespace-qualified (e.g. `OpenAPI\\Server\\Model\\Foo`), so it never matched `toModelName(...)` and the guard incorrectly fell into the "drop the default" branch for a genuine named-enum ref.

Pushed 2f8a35fb4f3, which swaps that check for `!languageSpecificPrimitives.contains(varDataType)` instead — a more direct test for "is there no enum class at this use site to qualify the value with", independent of per-language model-name formatting. Verified locally (JDK21):
- `PhpSymfonyServerCodegenTest#testModelPropertyEnumRefWithDefaultUsesNativeEnumCaseInCodegen` — now passes
- `JavaClientCodegenTest#testDiscriminatorPropertyRefToEnumDoesNotEmitInvalidDefault_issue24874` (the original #24874 regression test) — still passes
- `JavaModelEnumTest`, `DefaultCodegenTest`, `GoClientCodegenTest`, `DartClientCodegenTest`, `KotlinClientCodegenModelTest`, `KotlinClientCodegenApiTest`, `PythonClientCodegenTest`, `ConfluenceWikiTest`, `OpenApiEvaluatorTest` — all green, no regressions

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/java/JavaClientCodegenTest.java [hunk 1]
```diff
@@ -1869,6 +1869,27 @@ public void testObjectDefaultWithEnumProperty_issue24298() {
                 .contains("new OutputFormat().order(OutputFormat.OrderEnum.SIMILARITY).limit(10)");
     }
 
+    @Test
+    public void testDiscriminatorPropertyRefToEnumDoesNotEmitInvalidDefault_issue24874() {
```
Comment (jpfinne) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/java/JavaClientCodegenTest.java:1873]: Make it easier to write/review:

```
        Map<String, File> files = generateFromContract("src/test/resources/bugs/issue_24874.yaml", OKHTTP_GSON,
                Map.of(MODEL_NAME_PREFIX, "Stock"));
        JavaFileAssert.assertThat(files.get("StockArchiveCategoryEvent.java"))
                .fileDoesNotContain("String.ARCHIVE");
```

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java [hunk 2]
```diff
@@ -7155,7 +7155,25 @@ public void updateCodegenPropertyEnum(CodegenProperty var) {
                 }
             }
             if (enumName != null) {
-                var.defaultValue = toEnumDefaultValue(var, enumName);
```
Comment (jpfinne) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:7158]: Just update the invalid pojo.mustache.

Remove the line in 
https://github.com/OpenAPITools/openapi-generator/blob/c83886d9ebf2cba31007f4e05e4aec237eb10d68/modules/openapi-generator/src/main/resources/Java/libraries/okhttp-gson/pojo.mustache#L89

**Review by jpfinne — COMMENTED** [review]
updating DefaultCodegen is risky.

Just change the template

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/java/JavaClientCodegenTest.java (hunk 1 above)
Comment (seonwooj0810) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/JavaClientCodegenTest.java:1873]: Agreed, will switch this to generateFromContract + JavaFileAssert matching the other bugs/tests in this file, and push it in the next commit.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java (hunk 2 above)
Comment (seonwooj0810) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:7158]: I checked the git history on this line — it was added in #19614 (merged) specifically to fix enum-typed discriminators, and that PR closed five separate issues (#12412, #16073, #16672, #19194, #19261). Removing it unconditionally would regress that: any model where the discriminator property is genuinely typed as an enum would stop getting its default assigned in the subclass constructor.

#24874 is a narrower case — the discriminator is a $ref into another schema's property, not a genuinely-enum-typed discriminator — so I'd rather keep the DefaultCodegen-level guard and tighten it instead of removing the template line.

That said, your earlier concern was valid and exposed a real gap: the current languageSpecificPrimitives check doesn't cover mapped scalars like Date/UUID/BigDecimal, so a $ref into a property of one of those types would still emit an invalid qualified default (e.g. Date.CASE). I'll fix the check to key off whether the resolved schema is an actual named enum rather than primitive-set membership, which should also address cubic's P2 comment above.

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/java/JavaClientCodegenTest.java (hunk 1 above)
Comment (seonwooj0810) [reply modules/openapi-generator/src/test/java/org/openapitools/codegen/java/JavaClientCodegenTest.java:1873]: Done in 77718a38a3a — refactored to use `generateFromContract` + `JavaFileAssert`, matching the other bugs/tests in this file. Verified locally (JDK21, `-pl modules/openapi-generator -am test`) — passes.

**Comment by seonwooj0810** [comment]
Pushed 77718a38a3a for the test-style feedback (`generateFromContract` + `JavaFileAssert`, matching the rest of the file).

Re: "just update pojo.mustache" / "updating DefaultCodegen is risky" — I actually tried that route (drop the `this.{{name}} = {{defaultValue}};` line in okhttp-gson's pojo.mustache and let DefaultCodegen unconditionally compute the enum-qualified default) to see if it was lower-risk than tightening the DefaultCodegen guard. It isn't safe as a standalone change: that constructor line is what #19614 added so each discriminator subtype gets its own enum value initialized (e.g. `this.carType = CarType.SUV;`), and removing it unconditionally breaks `testEnumDiscriminatorDefaultValueIsNotString` — all 9 subtypes in that test stop getting their discriminator default assigned. So a template-only fix isn't actually simpler/safer here; happy to look at a version that keeps #19614 green, but the straightforward one doesn't.

Still outstanding from my earlier reply: tightening the DefaultCodegen check to key off the resolved schema being an actual named enum (rather than `languageSpecificPrimitives` membership), which should also close the mapped-scalar gap (Date/UUID/BigDecimal) cubic flagged. That's not in this push — I want to verify it against the #19614 fixture set (Cat/Dog/Gecko/Chameleon/MiniVan/CargoVan/SUV/Truck/Sedan) before landing it, will follow up separately rather than guess.

**Comment by jpfinne** [comment]
@seonwooj0810 you don't need to assign a value to the discriminator.  The annotation `@JsonSubTypes` is enough for correct serialization by Jackson.  All the other java libraries don't assign a value to the discriminator.

I think a good fix would be to add an import like `import org.openapi.model.CategorySource.CategoryEnum` for inline enums.
That would solve several generation issues.  For example 
src/test/resources/3_0/oneOfDiscriminator.yaml used in `SpringCodegenTest.testOneOfWithInheritedEnumDiscriminator()`
-> PetResponseEnumDisc.java misses an import on PetEnumDisc.PetTypeEnum
-> public PetTypeEnum getPetType() does not compile

Then the defaultValue could be set even for inline enums. The current template would generate `this.category = CategoryEnum.ARCHIVE;`

That fix is complex!  It requires changes in DiscriminatorCodegen, DiscriminationUtils, DiscriminatorData...

**Comment by seonwooj0810** [comment]
Thanks for the context. To be precise about what this PR actually does today: it only drops the default for the ref-to-property/primitive case that #24874 is about (e.g. a `String` discriminator that `$ref`s another schema's inline enum) — that's the `var.defaultValue = null` branch. For genuine named-enum discriminators (the #19614 fixture set — Cat/Dog/Gecko/Chameleon/SUV/Truck/Sedan/etc.) this PR still assigns the default, e.g. `this.carType = CarType.SUV;`, same as before.

If I'm reading your comment right, you're saying that assignment is unnecessary everywhere, not just in the broken ref-to-property case — since `@JsonSubTypes` handles serialization on its own and other java libraries don't assign it either. That's a bigger behavior change than what #24874 asked for: it would mean removing the `this.{{name}} = {{defaultValue}};` line from pojo.mustache unconditionally and updating `testEnumDiscriminatorDefaultValueIsNotString` (which currently asserts all 9 subtypes get that assignment) to match. That's the exact template-only route I held off on earlier, precisely because it changes #19614's behavior rather than just fixing the broken case.

Do you want me to go ahead and make that broader change (drop the assignment unconditionally + update the #19614 test), or keep this PR scoped to just the ref-to-property fix and treat "stop assigning discriminator defaults everywhere" as separate follow-up? Happy to do either, just want to confirm scope before changing #19614's established behavior.

Re: the import-based fix for inline enums (`CategorySource.CategoryEnum` etc.) — agreed that's a different, larger piece of work (DiscriminatorCodegen/DiscriminationUtils/DiscriminatorData) than this PR's scope. I'd rather not guess at that here; happy to see it filed as a separate issue if you'd like to track it.
